plugin/trino-druid/src/main/java/io/trino/plugin/druid/DruidJdbcClient.java

```diff
@@ -273,6 +302,70 @@ public WriteMapping toWriteMapping(ConnectorSession session, Type type)
         return legacyToWriteMapping(type);
     }
 
+    public static ColumnMapping timestampColumnMappingUsingSqlTimestampWithFullPushdown(TimestampType timestampType)
+    {
+        // Druid supports Timestamp with MILLI_SECOND Precision
+        checkArgument(timestampType.getPrecision() <= 3, "Precision is out of range: %s", timestampType.getPrecision());
+        return ColumnMapping.longMapping(
+                timestampType,
+                (resultSet, columnIndex) -> {
+                    // Druis's ResultSet depends on JDBC Connection TimeZone, so we pass the Calendar to get the result at UTC.
+                    Instant instant = Instant.ofEpochMilli(resultSet.getTimestamp(columnIndex, UTC_CALENDAR).getTime());
+                    LocalDateTime timestamp = LocalDateTime.ofInstant(instant, UTC);
+                    return toTrinoTimestamp(timestampType, timestamp);
+                },
+                timestampWriteFunctionUsingSqlTimestamp(timestampType),
+                /*
+                Druid's push down expression for "__time" i.e CAST('yyyy-mm-dd hh:MM:ss.zzz' AS TIMESTAMP) ignores the millisecond precision, so if the filter has a
+                millisecond precision then we would avoid pushdown.
+                 */
+                (session, domain) -> {
+                    boolean canPushdownFilter = domain.getValues().getValuesProcessor().transform(
+                            ranges -> ranges.getOrderedRanges().stream()
+                                    .allMatch(DruidJdbcClient::hasSecondPrecision),
+                            discreteValues -> {
+                                throw new UnsupportedOperationException("Not supported for discrete values");
+                            },
+                            allOrNone -> true);
+
+                    if (canPushdownFilter) {
+                        return FULL_PUSHDOWN.apply(session, domain);
+                    }
+                    return DISABLE_PUSHDOWN.apply(session, domain);
```

Comment:
Can we have mixed pushdown where try to pushdown (maybe broader) domain and keep the filter on Trino side?

Reply:
I was thinking it as an enhancement for the above PR. As we need special handling for `equality`, WDYT ?